data/transactions/logic/frames.go

```diff
@@ -0,0 +1,101 @@
+// Copyright (C) 2019-2022 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package logic
+
+import (
+	"errors"
+	"fmt"
+)
+
+func opProto(cx *EvalContext) error {
+	if !cx.fromCallsub {
+		return fmt.Errorf("proto executed in normal flow")
```

Comment:
No way to know what the last opcode was.  It might have been the previous opcode in cx.program, or someone might have `bnz`d to the `proto` instruction.  I'll say:
`return fmt.Errorf("proto was executed without a callsub")`
if you like that.